File: src/main/java/com/starshootercity/abilities/ParticleAbility.java
package com.starshootercity.abilities;

import com.destroystokyo.paper.event.server.ServerTickEndEvent;
import com.starshootercity.Origin;
import com.starshootercity.OriginSwapper;
import org.bukkit.Bukkit;
import org.bukkit.Particle;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;

public interface ParticleAbility extends Ability {
    Particle getParticle();
    int getFrequency();
    default int getExtra() {
        return 0;
    }
    default Object getData() {
        return null;
    }

    class ParticleAbilityListener implements Listener {
        @EventHandler
        public void onServerTickEnd(ServerTickEndEvent event) {
            for (Player player : Bukkit.getOnlinePlayers()) {
                Origin origin = OriginSwapper.getOrigin(player);
                if (origin == null) continue;
                for (Ability ability : origin.getAbilities()) {
                    if (ability instanceof ParticleAbility particleAbility) {
                        if (event.getTickNumber() % particleAbility.getFrequency() == 0) {
                            player.getWorld().spawnParticle(particleAbility.getParticle(), player.getLocation(), 1, 0.5, 1, 0.5, particleAbility.getExtra(), particleAbility.getData());
                        }
                    }
                }
            }
        }
    }
}

File: core/src/main/java/com/starshootercity/abilities/ParticleAbility.kt
package com.starshootercity.abilities

import com.destroystokyo.paper.event.server.ServerTickEndEvent
import com.starshootercity.OriginSwapper.Companion.getOrigins
import kotlinx.coroutines.runBlocking
import org.bukkit.Bukkit
import org.bukkit.Particle
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener

interface ParticleAbility : Ability {
    val particle: Particle
    val frequency: Int
        get() = 4
    val extra: Int
        get() = 0
    val data: Any?
        get() = null

    class ParticleAbilityListener : Listener {

        @EventHandler
        fun onServerTickEnd(event: ServerTickEndEvent) {
            runBlocking {
                Bukkit.getOnlinePlayers().forEach { player ->
                    getOrigins(player)
                        .flatMap { it.getAbilities() }
                        .filterIsInstance<ParticleAbility>()
                        .filter { event.tickNumber % it.frequency == 0 }
                        .forEach { ability ->
                            player.world.spawnParticle(
                                ability.particle,
                                player.location,
                                1,
                                0.5, 1.0, 0.5,
                                ability.extra.toDouble(),
                                ability.data
                            )
                        }
                }
            }
        }

        
    }
}
